Linux x64 - Cross-arch static compilation to x86 requires static_runtime=true

### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Kali Linux x64
| File format of the file you reverse (mandatory)      | n/a
| Architecture/bits of the file (mandatory)            | n/a
| `rizin -v` full output, **not truncated** (mandatory)  | n/a

### Expected behavior
Cross-arch compiling Rizin statically from x64 to x86 works when following the [BUILDING](https://github.com/rizinorg/rizin/blob/dev/BUILDING.md) guide (and forcing 32 bit build).

### Actual behavior
Cross-arch static compilation doesn't work without adding the `-Dstatic_runtime=true` meson flag.

### Steps to reproduce the behavior
```sh
$ git clone https://github.com/rizinorg/rizin
$ cd rizin
$ CFLAGS="-static -m32" LDFLAGS="-static -m32" meson --prefix="${PWD}/build/install" --default-library=static --buildtype=release build
$ ninja -C build
# 💔 - error: /usr/bin/ld: attempted static link of dynamic object `/usr/lib/gcc/x86_64-linux-gnu/10/../../../../lib32/libutil.so'
```

### Additional Logs, screenshots, source code,  configuration dump, ...
The full error is:
```
[...]
[1376/1550] Linking target binrz/rz-test/rz-test
FAILED: binrz/rz-test/rz-test 
cc  -o binrz/rz-test/rz-test binrz/rz-test/rz-test.p/rz-test.c.o binrz/rz-test/rz-test.p/load.c.o binrz/rz-test/rz-test.p/run.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,-O1 -static -m32 -static -m32 -Wl,--start-group librz/util/librz_util.a librzspp.a subprojects/sdb/src/libsdb.a subprojects/zlib-1.2.11/libz.a -ldl -lm -pthread /usr/lib/gcc/x86_64-linux-gnu/10/../../../../lib32/libutil.so -Wl,--end-group
/usr/bin/ld: librz/util/librz_util.a(lib.c.o): in function `rz_lib_dl_open':
lib.c:(.text+0xa0): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/bin/ld: attempted static link of dynamic object `/usr/lib/gcc/x86_64-linux-gnu/10/../../../../lib32/libutil.so'
collect2: error: ld returned 1 exit status
[...]
```

**Important**: This error does **not** happen (meaning that the cross compilation is successful) when passing `-Dstatic_runtime=true` to meson:

```sh
$ git clone https://github.com/rizinorg/rizin
$ cd rizin
$ CFLAGS="-static -m32" LDFLAGS="-static -m32" meson --prefix="${PWD}/build/install" --default-library=static -Dstatic_runtime=true --buildtype=release build
$ ninja -C build
# ❤️ - Though with many warnings in the linking process saying:
# /usr/bin/ld: librz/util/librz_util.a(lib.c.o): in function `rz_lib_dl_open':
# lib.c:(.text+0xa0): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```

Currently the `--default-library=static` meson flag doesn't seem to imply `-Dstatic_runtime=true` according to [this](https://github.com/rizinorg/rizin/blob/dev/meson.build#L20) (and the following) lines, but I don't know if this is wanted behavior.

Note that I purposely used Kali Linux as it is a popular distro in the InfoSec world, install `gcc-multilib` for being able to cross-compile.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -365,7 +365,7 @@ jobs:
     - name: Compile with meson
       run: |
         mkdir -p ${HOME}/rizin-static
-        CFLAGS="-static" LDFLAGS="-static" meson --prefix=${HOME}/rizin-static --buildtype release --default-library static -Dstatic_runtime=true -Dportable=true build
+        meson --prefix=${HOME}/rizin-static --buildtype release --default-library static -Dstatic_runtime=true -Dportable=true build
         ninja -C build && ninja -C build install
         tar -C ${HOME}/rizin-static --xz -cf rizin-static.tar.xz $(ls ${HOME}/rizin-static)
       working-directory: rizin
@@ -550,7 +550,7 @@ jobs:
     - name: Compile with meson
       run: |
         export PATH=${HOME}/.local/bin:${PATH}
-        CFLAGS="-static" LDFLAGS="-static" meson --buildtype release --default-library static --prefix=/tmp/android-dir -Dportable=true -Dblob=true build --cross-file .github/meson-android-${{ matrix.name }}.ini
+        meson --buildtype release --default-library static --prefix=/tmp/android-dir -Dportable=true -Dblob=true -Dstatic_runtime=true build --cross-file .github/meson-android-${{ matrix.name }}.ini
         ninja -C build && ninja -C build install
     - name: Create rizin-android-${{ matrix.name }}.tar.gz
       run: |

--- a/BUILDING.md
+++ b/BUILDING.md
@@ -104,11 +104,12 @@ $ meson -Db_sanitize=address,undefined build
 
 ## Build fully-static binaries
 
-It may be useful to run Rizin just by using a single file, which can be copied
-on other systems if necessary.
+It may be useful to run Rizin just by using a single file, which can be
+copied on other systems if necessary. On *NIX systems, this adds the classic
+`-static` flag to the linker, while on Windows it uses `/MT`.
 
 ```
-$ CFLAGS="-static" meson --default-library=static build
+$ meson --default-library=static -Dstatic_runtime=true build
 ```
 
 ## Cross-compilation for Android
@@ -131,13 +132,72 @@ feature, which will produce just one executable and link all the other tools to
 that only tool, similar to how busybox works.
 
 ```
-$ CFLAGS="-static" LDFLAGS="-static" meson --buildtype release --default-library static --prefix=/tmp/android-dir -Dblob=true build --cross-file ./cross-compile-conf.ini
+$ meson --buildtype release --default-library static --prefix=/tmp/android-dir -Dblob=true build -Dstatic_runtime=true --cross-file ./cross-compile-conf.ini
 $ ninja -C build
 $ ninja -C build install
 ```
 
 At this point you can find everything under `/tmp/android-dir` and you can copy
 files to your Android device.
+
+## Compile 32-bit Rizin on 64-bit machine
+
+Whenever you want to build Rizin for a different system/architecture than the
+one you are using for building, you are effectively cross-compiling and you
+should provide a full configuration file to tell meson what is the target
+machine.
+
+Even to compile a 32-bit version of Rizin on a 64-bit machine, you should use
+a configuration file like the following:
+
+```
+[binaries]
+c = '/usr/bin/gcc'
+cpp = '/usr/bin/g++'
+ar = '/usr/bin/gcc-ar'
+strip = '/usr/bin/strip'
+pkgconfig = '/usr/bin/i686-redhat-linux-gnu-pkg-config'
+llvm-config = '/usr/bin/llvm-config-32'
+
+[built-in options]
+c_args = ['-m32']
+c_link_args = ['-m32']
+cpp_args = ['-m32']
+cpp_link_args = ['-m32']
+
+[host_machine]
+system = 'linux'
+cpu_family = 'x86'
+cpu = 'i686'
+endian = 'little'
+```
+
+Alternatively, if your distribution provide specific compiler tools for the
+i686 architecture, you can use a configuration similar to this:
+
+```
+[binaries]
+c = '/usr/bin/i686-linux-gnu-gcc'
+cpp = '/usr/bin/i686-linux-gnu-g++'
+ar = '/usr/bin/i686-linux-gnu-gcc-ar'
+strip = '/usr/bin/i686-linux-gnu-strip'
+pkgconfig = '/usr/bin/i686-linux-gnu-pkg-config'
+
+[host_machine]
+system = 'linux'
+cpu_family = 'x86'
+cpu = 'i686'
+endian = 'little'
+```
+
+Of course, you might have to adjust some settings depending on your system
+and you should double check that you have installed all the necessary 32-bit
+libraries and tools. Once you have checked everything, you can setup the
+build directory with:
+
+```
+$ meson build --cross-file ./rizin-i386.ini
+```
 
 # Uninstall
 

--- a/meson.build
+++ b/meson.build
@@ -109,9 +109,18 @@ if host_machine.system() == 'windows'
 endif
 platform_inc = include_directories(platform_inc)
 
-if is_static_build and cc.get_id() == 'msvc'
-  # Use -Db_vscrt=static_from_buildtype to avoid warnings due to multiple /M options
-  add_project_arguments('/MT', language: 'c')
+if is_static_build
+  if cc.get_id() == 'msvc'
+    # Use -Db_vscrt=static_from_buildtype to avoid warnings due to multiple /M options
+    if get_option('b_vscrt') == 'from_buildtype'
+      warning('To avoid warnings due to multiple /M options, please also set -Db_vscrt=mt or -Db_vscrt=mtd.')
+      add_project_arguments('/MT', language: 'c')
+    endif
+  else
+    if cc.has_link_argument('-static')
+      add_project_link_arguments('-static', language: 'c')
+    endif
+  endif
 endif
 
 if cc.has_argument('--std=gnu99')

